src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::context::{self, Context};
+use crate::events::EventNotifier;
+use crate::worker::{self, Worker};
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Order {
+    LeftToRight,
+    RightToLeft,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Strategy {
+    OneForOne,
+}
+
+#[derive(Debug)]
+pub struct StartError {
+    pub failing_child_runtime_name: String,
+    pub failing_child_error: Arc<worker::StartError>,
+    pub failing_sibling_termination_error: Option<TerminationError>,
+}
+
+#[derive(Debug)]
+pub struct TerminationError(HashMap<String, Arc<worker::TerminationError>>);
+
+pub struct Spec {
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+    pub children: Vec<worker::Spec>,
+}
+
+pub struct Supervisor {
+    pub runtime_name: String,
+    pub children: Vec<Worker>,
+
+    pub name: String,
+    pub ev_notifier: EventNotifier,
+    pub order: Order,
+    pub strategy: Strategy,
+}
```

Comment:
Kind of a quibble, but these seem to have a lot of the same fields, which I suspect is intentional. If I'm understanding correctly, a `Spec` is the finished/cancelled format of the same data in a `Supervisor` which holds the data while it is running.  I wonder if there would be any utility in having the `Spec` or some other struct be the set of common fields with two different containers?

Another possibility (and probably the most idomatic Rust way), since the only difference is actually the type of the children, would be to do something like:

```suggestion
pub struct SpecBase<Child> {
    pub name: String,
    pub ev_notifier: EventNotifier,
    pub order: Order,
    pub strategy: Strategy,
    pub children: Vec<Child>,
}

pub type Spec = SpecBase<worker::Spec>;
pub type Supervisor = SpecBase<Worker>;
```

Reply:
That is pretty neat pattern, that being said, I'm certain `Supervisor` will have more attributes than `Spec` in a more complete implementation. 

Given this, wouldn't having the same `SpecBase` for both `Spec` and `Supervisor` make the evolution of `Supervisor` more difficult in the long run?

Reply:
Ah, definitely true. If you know there will be more fields definitely leave this as is.

Reply:
I went with a `SpecMeta` that contains all the common data